### PR TITLE
perf(api): Move group's current release to its own endpoint

### PR DESCRIPTION
**Pull request by RyanSkonnord**
As a performance fix, "currentRelease" is no longer part of the issue
details view, but is provided by its own endpoint. The intent is that
the new endpoint would be called later in the page load, in preference
to blocking issue details with a costly database query.

Also, change the API contract such that, if no environment parameters
are provided, search across all environments associated with the
group's project for the latest release. Previously, the group details
endpoint would omit "currentRelease" entirely when not filtering by
environment. We can afford to do it in that most common case, now that
it is not a performance concern. Note the distinction from returning a
null "currentRelease" value, which still happens under the same
conditions as before.

### CODE REVIEW

**Comment by RyanSkonnord** [comment]
Per @leedongwei's suggestion to solve the performance problem on page load by breaking "current release" from issue details into its own endpoint (and also to include all environments when the user is not filtering by any).

Can't be merged as is, because removing it from the issue details endpoint would simply break the feature until the front end has been updated to make a delayed call to the new endpoint. Both changes should be deployed at once (assuming that fixing the performance problem isn't massively more important than preserving the feature).

File: src/sentry/api/endpoints/group_current_release.py
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/group_current_release.py:47]: As long as we're messing with the API contract, I wouldn't mind changing the `"currentRelease": null` thing into something more meaningful as well.

Perhaps
```javascript
{
  "currentRelease": { /* always points at the actual release unless none exists */ },
  "issueSeenInCurrentRelease": false
}
```

Only tricky part is that we serialize a `GroupRelease` object to populate `"currentRelease"`, which of course doesn't exist in the "false" case. Maybe I'll find a good option if I play around with the `ReleaseEnvironment` model. It will probably make the database interaction more expensive overall, but luckily we don't need to care any more! 😜

**Comment by leedongwei** [comment]
@RyanSkonnord I've updated the frontend to use this endpoint on commit 03d1f9bc.

I made that commit on top of #20168, so that PR will have to be merged and we'll have to rebase onto the `master` before I can merge that commit in.

File: src/sentry/static/sentry/app/components/group/releaseStats.tsx
Comment (billyvg) [inline src/sentry/static/sentry/app/components/group/releaseStats.tsx:16]: Can we go ahead and type this since you're actively working/testing it?

File: src/sentry/static/sentry/app/components/group/releaseStats.tsx
Comment (billyvg) [inline src/sentry/static/sentry/app/components/group/releaseStats.tsx:28]: Is there a reason for changing this?

File: src/sentry/static/sentry/app/components/group/sidebar.tsx
Comment (billyvg) [inline src/sentry/static/sentry/app/components/group/sidebar.tsx:74]: Can we refactor these to use `api.requestPromise` while you're in there? (It's also fine to leave as-is)

File: src/sentry/static/sentry/app/components/group/releaseStats.tsx
Comment (leedongwei) [reply src/sentry/static/sentry/app/components/group/releaseStats.tsx:28]: Was tweaking the entire component around but I undid that. No reason to change it now so reverting back.

File: src/sentry/static/sentry/app/components/group/releaseStats.tsx
Comment (leedongwei) [reply src/sentry/static/sentry/app/components/group/releaseStats.tsx:16]: I'll follow up in a separate PR for this.

**Review by billyvg — COMMENTED** [review]
The UI code LGTM - I have not manually tested it, do we have good loading placeholders when currentRelease is loading?

Looks like our current UI code handles the case when `group.currentRelease` doesn't exist, so we don't have to worry about old clients making requests to the new API.

Is there a reason why we don't make the currentRelease request inside of the component that needs it? That would probably save on a re-render for the sidebar.

**Review by billyvg — COMMENTED** [review]
The UI code LGTM - I have not manually tested it, do we have good loading placeholders when currentRelease is loading?

Looks like our current UI code handles the case when `group.currentRelease` doesn't exist, so we don't have to worry about old clients making requests to the new API.

Is there a reason why we don't make the currentRelease request inside of the component that needs it? That would probably save on a re-render for the sidebar.

File: src/sentry/static/sentry/app/components/group/releaseStats.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/group/releaseStats.tsx:16]: any | null doesn't make much sense, but for now, you could write it as below

```suggestion
  currentRelease?: any | null  // TODO(ts)
```

File: src/sentry/api/endpoints/group_details.py
Comment (markstory) [inline src/sentry/api/endpoints/group_details.py:322]: Issue details is a documented API endpoint. You might end up breaking external integrations with this removal. cc @getsentry/ecosystem

Comment (RyanSkonnord) [reply src/sentry/api/endpoints/group_details.py:324]: We did investigate that at the time and determined that `currentRelease` is internal-only, or at least that seems to be the intent based on the code history and all available documentation. If it were being consumed externally, I would expect the rather obscure meaning of null/non-null values to be explained somewhere.

But, 👍 to getting approval from @getsentry/ecosystem.

Comment (scefali) [reply src/sentry/api/endpoints/group_details.py:324]: @RyanSkonnord Nah, Ecosystem does not even use the GET method on this endpoint. We do stuff for updating the issue/group to be resolved in the current release but that's a PUT request.

File: src/sentry/api/endpoints/group_current_release.py
Comment (RyanSkonnord) [reply src/sentry/api/endpoints/group_current_release.py:16]: In case it's not obvious, this is the same code that was removed from `group_details.py`, with a slight refactor for readability but functionally the same query.

I tried several other ways of optimizing this but I don't think any of them used `Subquery`. Some of my experiments involved hand-tuned raw SQL queries, some of which would I think do something similar.

It sounds like a good idea, but I would rather kick it forward to a separate investigation task.

Comment (markstory) [reply src/sentry/api/endpoints/group_current_release.py:16]: Doing it separately sounds good to me. The old code was likely not using Subquery() as we were likely on old django that didn't support that behavior yet.

**Comment by leedongwei** [comment]
Might break due to #21423

File: src/sentry/static/sentry/app/components/group/releaseStats.tsx
Comment (leedongwei) [reply src/sentry/static/sentry/app/components/group/releaseStats.tsx:18]: Updated with types in this PR.

**Comment by leedongwei** [comment]
> The UI code LGTM - I have not manually tested it, do we have good loading placeholders when currentRelease is loading?

It's displaying a number in the graph's tooltip  `events in release 4a509...: 0` so setting loading state for that would be tricky. If it's loading, it'll just not show anything.

Graph and tooltip in question:


I'm not sure if this would be a good solution:


<br>

> Is there a reason why we don't make the currentRelease request inside of the component that needs it? That would probably save on a re-render for the sidebar.

I like the simplicity/centralization of `GroupSidebar` managing all the fetching.

**Review by billyvg — COMMENTED** [review]
Is this still being worked on? UI code still gets a 👍 from me, will leave it up for backend approval.

**Review by leedongwei — APPROVED** [review]

**Pull request merged**
